pkg/controller/master-controller-manager/application-secret-synchronizer/controller.go

```diff
@@ -0,0 +1,172 @@
+		masterClient: masterManager.GetClient(),
+		seedClients:  map[string]ctrlruntimeclient.Client{},
+	}
+
+	for seedName, seedManager := range seedManagers {
+		r.seedClients[seedName] = seedManager.GetClient()
+	}
+
+	c, err := controller.New(ControllerName, masterManager, controller.Options{Reconciler: r, MaxConcurrentReconciles: numWorkers})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %w", err)
+	}
+
+	if err := c.Watch(&source.Kind{Type: &k8scorev1.Secret{}}, &handler.EnqueueRequestForObject{}, predicate.ByAnnotation(secretTypeAnnotation, "", false)); err != nil {
+		return fmt.Errorf("failed to create watch for secrets: %w", err)
+	}
+
+	return nil
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("resource", request.Name)
+	log.Debug("Processing")
+
+	err := r.reconcile(ctx, log, request)
+	if err != nil {
+		log.Errorw("ReconcilingError", zap.Error(err))
+	}
+
+	return reconcile.Result{}, err
+}
+
+func (r *reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, request reconcile.Request) error {
+	secret := &k8scorev1.Secret{}
+
+	var err error
+	if err = r.masterClient.Get(ctx, request.NamespacedName, secret); err != nil {
+		if !kerrors.IsNotFound(err) {
+			return err
+		} else {
+			// handling deletion
+			delSecret := &k8scorev1.Secret{ObjectMeta: v1.ObjectMeta{Name: request.Name, Namespace: seedNS}}
+			if err := r.handleDeletion(ctx, log, delSecret); err != nil {
+				return fmt.Errorf("failed to delete secret: %w", err)
+			}
+			return nil
+		}
+	}
+
+	seedsecret := secret.DeepCopy()
+	seedsecret.Namespace = seedNS
+	seedsecret.SetResourceVersion("")
+
+	namedSecretCreatorGetter := []reconciling.NamedSecretCreatorGetter{
+		secretCreator(seedsecret),
+	}
+	err = r.reconcileAllSeeds(ctx, log, seedsecret, func(ctx context.Context, c ctrlruntimeclient.Client, o ctrlruntimeclient.Object) error {
+		return reconciling.ReconcileSecrets(ctx, namedSecretCreatorGetter, seedNS, c)
+	})
+	if err != nil {
+		r.recorder.Eventf(secret, k8scorev1.EventTypeWarning, "ReconcilingError", err.Error())
+		return fmt.Errorf("reconciling secret %s failed: %w", seedsecret.Name, err)
+	}
+
+	return nil
+}
+
+func secretCreator(s *k8scorev1.Secret) reconciling.NamedSecretCreatorGetter {
+	return func() (name string, create reconciling.SecretCreator) {
+		return s.Name, func(existing *k8scorev1.Secret) (*k8scorev1.Secret, error) {
+			return s, nil
+		}
+	}
+}
+
+func (r *reconciler) handleDeletion(ctx context.Context, log *zap.SugaredLogger, secret *k8scorev1.Secret) error {
+	delfunc := func(ctx context.Context, c ctrlruntimeclient.Client, o ctrlruntimeclient.Object) error {
+		err := c.Get(ctx, types.NamespacedName{Name: secret.Name, Namespace: secret.Namespace}, &k8scorev1.Secret{})
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				log.Infof("Secret %q already deleted", secret.Name)
```

Comment:
I am not sure I full understand. So I mean this already produces a structured message:

```
{"level":"info","time":"2022-04-28T12:21:08.719+0200","caller":"application-secret-synchronizer/controller.go:143","msg":"Secret \"secret-1\" already deleted","resource":"secret-1”}
```

Or do you want here that just the logger is extended to have the secret name as a structured field and not templated into the message? Something like this:

```Go
log := log.With("secret", secret.Name)
log.Infof("secret was already deleted")
```

```
{"level":"info","time":"2022-04-28T18:22:43.084+0200","caller":"application-secret-synchronizer/controller.go:144","msg":"secret was already deleted","resource":"secret-1","secret":"secret-1"}
```

If so: do we then generally prefer not to log template data in the `Errorf` section, but rather extract them all into separate fields? (asking because then I will make a pull-request to https://github.com/kubermatic/kubermatic/blob/master/docs/logging.md to clarify)

Reply:
```go
log := log.With("secret", secret.Name)
log.Info("secret was already deleted")
```

This is the way, correct. In general we try to avoid formatting messages (though sometimes, just for readability, it makes sense). However controllers usually put the object they are reconciling as a context field in the logger, because reconcilers usually produce multiple log messages. Having them all be tagged properly is nice.